File: primalityTest/Sieve of Eratosthenes/SieveofEratosthenes.cpp
/*The Sieve of Eratosthenes finds its application in generating all prime numbers up to a particular number. The 
prime numbers do not have factors other than one and themselves. We try to exploit the property in reverse order
 i.e. number having more than 2 factors is not a prime number.
 We follow the given steps to do this.
 1.Assume all numbers are prime.
 2.Mark 0 and 1 as non primes.
 3.From 2 to n we iterare and mark all the multiples of these numbers are non prime.(As they have atleast factors now(1,themselves and this number))
 4.For easing out our task we are not going for multiples of numbers marked as non primes as they would have been alraedy dealt by its factors.
    for example: 2 will knock down 4,6,8,10,12,14,16,18,20.
    So we don't need to check for 4 as 2 has already knocked down its multiples.
 5.At the end we will be left with only prime values that are marked as prime.
 ***************************************************************************************************************************************************
 TIME COMPLEXITY: O(N*logN*logN)                          for proof refer this:https://cp-algorithms.com/algebra/sieve-of-eratosthenes.html
 SPACE COMPLEXITY:O(N)
 ***************************************************************************************************************************************************
*/
#include<bits/stdc++.h>
using namespace std;

int main()
{
    int n=100;                                             // defining higher limit, generate prime numbers in range[1:100]
    vector<bool> checkprime(n+1,true);                     // assuming all the numbers are prime numbers
    vector<int> prime;                                     // vector for storing all the prime numbers
    checkprime[0]=false;                                   // knocking out 0 and 1 from the list of prime numbers
    checkprime[1]=false;                                   // starting our search for prime numbers from [2:n]
    for(int i=2;i<=n;i++)
    {
        if(checkprime[i]==true)                             // We will check only those which are not knocked out yet.(POINT 4)
        {
            for(int j=2*i;j<=n;j+=i)                        // this will generate all the multiples of i upto n. 2*i will be the smallest multiple.
            {
                checkprime[j]=false;                        // knocking out the multiples from pt=rime list
            }
        }
    }                              
    for(int i=0;i<n+1;i++)                                   // iterating over checkprime vector
    {
        if(checkprime[i]==true)                              // copying the prime number in prime.
        prime.push_back(i);
    }
    return 0;
}

File: primalityTest/Sieve of Eratosthenes/sieveOfEratosthenes.java
/*
In mathematics, the sieve of Eratosthenes is an ancient algorithm for finding all prime numbers up to any given limit.

It does so by iteratively marking as composite (i.e., not prime) the multiples of each prime, starting with the first prime number, 2. 
The multiples of a given prime are generated as a sequence of numbers starting from that prime, with constant difference between them 
that is equal to that prime.[1] This is the sieve's key distinction from using trial division to sequentially test each candidate number 
for divisibility by each prime.[2] Once all the multiples of each discovered prime have been marked as composites, the remaining unmarked
numbers are primes.

Algorithm:
1) Create a list of consecutive integers from 2 through n: (2, 3, 4, ..., n).
2) Initially, let p equal 2, the smallest prime number.
3) Enumerate the multiples of p by counting in increments of p from 2p to n, and mark them in the list (these will be 2p, 3p, 4p, ...; 
   the p itself should not be marked).
4) Find the smallest number in the list greater than p that is not marked. If there was no such number, stop. Otherwise, let p now equal 
   this new number (which is the next prime), and repeat from step 3.
5) When the algorithm terminates, the numbers remaining not marked in the list are all the primes below n.

Source : https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
*/ 


import java.util.*;

class driver{
    public static void main(String[] args) {
        int MAX = 1000;                                    // setting upper limit => Sieve generates prime numbers in [1, MAX]

        boolean isPrime[] = new boolean[MAX + 1];          // List of numbers 
        Arrays.fill(isPrime, true);                        // assuming all numbers to be prime

        isPrime[0] = false;                                // 0 is not a prime number hence eliminating it at first
        isPrime[1] = false;                                // 1 is not a prime number hence eliminating it at first

        for(int i = 2; i * i <= MAX; i++) {                // search space is from [2 : sqrt(MAX)]    => where sqrt = square root
            if(isPrime[i]) {
                for(int j = i * i; j <= MAX; j += i) {     // this generates all multiples of i starting from i*i. All previous multiples of i have already been checked
                    isPrime[j] = false;                    // removes the multiples of i from the prime list
                }
            }
        }

        for(int i = 0; i <= MAX; i++) {                    // iterating over isPrime vector
            if(isPrime[i]) {                               // if isPrime[i] is true, it means that i is a prime number
                System.out.println(i);                     // prints i
            }
        }

    }
}

/*
TIme Complexity : O(nlog(logn))
Space Complexity : O(n)
*/